> ### The Battle of Neighborhoods :
> ---------------------------------------------------------------
> # Comparing Neighborhoods in The Bronx and Brooklyn 
-----------------------------------------------------------------

##### -  Table of Content  -

Part 1. Introduction
Part 2. Data
3. Methodology
4. Result
5. Conclusion & Discussion

## Part 1: Introduction
### 1.1 Description of the Problem

Brooklyn is on the rise while The Bronx seems unfortunately in decline.  

One of my collegue has to move into the Bronx who have been in Brooklyn for a long time.  For my collegue who need to choose new neighborhood in the new city, I make up my mind to give a real insight between the two cities: Brooklyn VS the Bronx.    

Also, Brooklyn and The Bronx have their own unique characteristic.  I believe that exploring and comparing these two cities would be interesting.   
By discovering how similar or dissimilar they are, I will give a real insights for my colleague.

### 1.2 Background
Brooklyn and The Bronx are very different. They don't mix like oil and water.  

One reason is that Staten Island aside, which is totally separate and apart, The Bronx and Brooklyn are the two most geographically distant boroughs. Just as a matter of time and distance, residents of the two boroughs rarely mix.

The basic difference is that Brooklyn and The Bronx developed differently. Brooklyn has its own separate and independant identity as a city. For much of its existence as one of the largest in the country. Even as a borough, it would qualify as the 3rd or 4th largest city in the USA.

The city of Brooklyn developed in parralell to the City of New York, rivaling NYC until they merged, evidenced by Brooklyn's downtown and Brownstone Belt. Also, the outlying areas of Brooklyn (before Brooklyn was consolidated) were well developed and prosperous, such as Flatbush, Bushwhich, and the other Dutch towns. 

The Bronx was an outlying area of mostly farmland and remained this way for most of its existence, until the subway was extended to the south bronx area. Afterwhich, The Bronx, unlike Brooklyn, developed with dense neighborhoods of mostly apartment buildings in which the masses of the lower east side expunged upon.

### 1.3 Target Audience
For this research, my target audience is my colleague. However, these kinds of researches will provide insights for the clients who need to decide a correct location.    
For example, as a data scientist for my client, I can recommend to the marketing team which neighborhood will be the best choice for maximizing their marketing activity performance.

In [1]:
import numpy as np 
import pandas as pd
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes
import folium # map rendering library

print('Libraries imported.')

Solving environment: done

# All requested packages already installed.

Libraries imported.


## Part 2. Data
### Data 1 
- **new_york_dataset**: I downloaded a json file. This json contains all the names, the borough, and latitude and longitude of every neighborhood in New York City. I extracted the json file from the internet.

### Data 2
- **Foursquare location API**: I got the top 100 venues within a 750 meter radius of each neighborhood by using the Foursquare location API.
---------------------------------------------------------------------------

#### The code below is how I extracted the json file from the internet: 

In [2]:
!wget -q -O 'newyork_data.json' https://cocl.us/new_york_dataset
print('Data downloaded!')

Data downloaded!


In [3]:
# Load the data.
with open('newyork_data.json') as json_data:
    newyork_data = json.load(json_data)

In [4]:
# Take a look at the data.
newyork_data

{'type': 'FeatureCollection',
 'totalFeatures': 306,
 'features': [{'type': 'Feature',
   'id': 'nyu_2451_34572.1',
   'geometry': {'type': 'Point',
    'coordinates': [-73.84720052054902, 40.89470517661]},
   'geometry_name': 'geom',
   'properties': {'name': 'Wakefield',
    'stacked': 1,
    'annoline1': 'Wakefield',
    'annoline2': None,
    'annoline3': None,
    'annoangle': 0.0,
    'borough': 'Bronx',
    'bbox': [-73.84720052054902,
     40.89470517661,
     -73.84720052054902,
     40.89470517661]}},
  {'type': 'Feature',
   'id': 'nyu_2451_34572.2',
   'geometry': {'type': 'Point',
    'coordinates': [-73.82993910812398, 40.87429419303012]},
   'geometry_name': 'geom',
   'properties': {'name': 'Co-op City',
    'stacked': 2,
    'annoline1': 'Co-op',
    'annoline2': 'City',
    'annoline3': None,
    'annoangle': 0.0,
    'borough': 'Bronx',
    'bbox': [-73.82993910812398,
     40.87429419303012,
     -73.82993910812398,
     40.87429419303012]}},
  {'type': 'Feature',
 

#### All of the relevant data is in the features key. Define a new variable that includes the features data.

In [5]:
neighborhood_data = newyork_data['features']

In [6]:
# Take a look at the first item in the list.
neighborhood_data[0]

{'type': 'Feature',
 'id': 'nyu_2451_34572.1',
 'geometry': {'type': 'Point',
  'coordinates': [-73.84720052054902, 40.89470517661]},
 'geometry_name': 'geom',
 'properties': {'name': 'Wakefield',
  'stacked': 1,
  'annoline1': 'Wakefield',
  'annoline2': None,
  'annoline3': None,
  'annoangle': 0.0,
  'borough': 'Bronx',
  'bbox': [-73.84720052054902,
   40.89470517661,
   -73.84720052054902,
   40.89470517661]}}

#### Transform the data into a pandas dataframe.
transform the data from nested Python dictionaries into a pandas dataframe.

In [7]:
# Define the dataframe columns.
column_names = ['Borough', 'Neighborhood', 'Latitude', 'Longitude']

neighborhoods = pd.DataFrame(columns=column_names)

In [8]:
neighborhoods

Empty DataFrame
Columns: [Borough, Neighborhood, Latitude, Longitude]
Index: []

#### Loop through the data and fill the dataframe.
The feature key contains the four items I need for my dataframe (Borough, Neighborhood, Latitude and Longitude). Next I created a blank Pandas data frame, and then looped the json data into the new Pandas data frame using the for loop below:

In [9]:
for data in neighborhood_data:
    borough = neighborhood_name = data['properties']['borough'] 
    neighborhood_name = data['properties']['name']
        
    neighborhood_latlon = data['geometry']['coordinates']
    neighborhood_lat = neighborhood_latlon[1]
    neighborhood_lon = neighborhood_latlon[0]
    
    neighborhoods = neighborhoods.append({'Borough': borough,
                                    'Neighborhood': neighborhood_name,
                                    'Latitude': neighborhood_lat,
                                    'Longitude': neighborhood_lon}, ignore_index=True)

In [10]:
neighborhoods.head()

Borough Neighborhood   Latitude  Longitude
0   Bronx    Wakefield  40.894705 -73.847201
1   Bronx   Co-op City  40.874294 -73.829939
2   Bronx  Eastchester  40.887556 -73.827806
3   Bronx    Fieldston  40.895437 -73.905643
4   Bronx    Riverdale  40.890834 -73.912585

In [11]:
# Change the name of the dataframe.
bxbk_df = neighborhoods

In [12]:
bxbk_df.head()

Borough Neighborhood   Latitude  Longitude
0   Bronx    Wakefield  40.894705 -73.847201
1   Bronx   Co-op City  40.874294 -73.829939
2   Bronx  Eastchester  40.887556 -73.827806
3   Bronx    Fieldston  40.895437 -73.905643
4   Bronx    Riverdale  40.890834 -73.912585

Now that my data is in a Pandas dataframe, it'll be much easier to manipulate and examine. Since I am only interested in neighborhoods in The Bronx and Brooklyn, I simply dropped all the rows that contained neighborhoods in Manhattan, Queens and Staten Island. Yes, Staten Island is still technically a borough of NYC. I know it's easy to forget about. Next I needed to get the top 100 venues within a 750 meter radius of each neighborhood using the Foursquare location API. I used the code below to do that:

In [13]:
# I only want to look at neighborhoods in the Bronx and Brooklyn. 
# Drop rows with Manhattan, Queens, and Staten Island.
bxbk_df.drop(neighborhoods.index[bxbk_df['Borough'] == 'Manhattan'], inplace = True)
bxbk_df.drop(neighborhoods.index[bxbk_df['Borough'] == 'Queens'], inplace = True)
bxbk_df.drop(neighborhoods.index[bxbk_df['Borough'] == 'Staten Island'], inplace = True)

In [14]:
# Check to make sure  Bronx and Brooklyn are left in the dataframe.
bxbk_df.Borough.unique()

array(['Bronx', 'Brooklyn'], dtype=object)

In [15]:
bxbk_df.reset_index(drop=True)

Borough               Neighborhood   Latitude  Longitude
0       Bronx                  Wakefield  40.894705 -73.847201
1       Bronx                 Co-op City  40.874294 -73.829939
2       Bronx                Eastchester  40.887556 -73.827806
3       Bronx                  Fieldston  40.895437 -73.905643
4       Bronx                  Riverdale  40.890834 -73.912585
5       Bronx                Kingsbridge  40.881687 -73.902818
6       Bronx                   Woodlawn  40.898273 -73.867315
7       Bronx                    Norwood  40.877224 -73.879391
8       Bronx             Williamsbridge  40.881039 -73.857446
9       Bronx                 Baychester  40.866858 -73.835798
10      Bronx             Pelham Parkway  40.857413 -73.854756
11      Bronx                City Island  40.847247 -73.786488
12      Bronx               Bedford Park  40.870185 -73.885512
13      Bronx         University Heights  40.855727 -73.910416
14      Bronx             Morris Heights  40.847898 -73.919672
15      Bronx                    Fordham  40.860997 -73.896427
16      Bronx               East Tremont  40.842696 -73.887356
17      Bronx                 West Farms  40.839475 -73.877745
18      Bronx               High  Bridge  40.836623 -73.926102
19      Bronx                    Melrose  40.819754 -73.909422
20      Bronx                 Mott Haven  40.806239 -73.916100
21      Bronx                Port Morris  40.801664 -73.913221
22      Bronx                   Longwood  40.815099 -73.895788
23      Bronx                Hunts Point  40.809730 -73.883315
24      Bronx                 Morrisania  40.823592 -73.901506
25      Bronx                  Soundview  40.821012 -73.865746
26      Bronx               Clason Point  40.806551 -73.854144
27      Bronx                Throgs Neck  40.815109 -73.816350
28      Bronx               Country Club  40.844246 -73.824099
29      Bronx                Parkchester  40.837938 -73.856003
30      Bronx         Westchester Square  40.840619 -73.842194
31      Bronx                   Van Nest  40.843608 -73.866299
32      Bronx                Morris Park  40.847549 -73.850402
33      Bronx                    Belmont  40.857277 -73.888452
34      Bronx             Spuyten Duyvil  40.881395 -73.917190
35      Bronx            North Riverdale  40.908543 -73.904531
36      Bronx                 Pelham Bay  40.850641 -73.832074
37      Bronx              Schuylerville  40.826580 -73.826203
38      Bronx             Edgewater Park  40.821986 -73.813885
39      Bronx                Castle Hill  40.819014 -73.848027
40      Bronx                  Olinville  40.871371 -73.863324
41      Bronx             Pelham Gardens  40.862966 -73.841612
42      Bronx                  Concourse  40.834284 -73.915589
43      Bronx                  Unionport  40.829774 -73.850535
44      Bronx                   Edenwald  40.884561 -73.848083
45   Brooklyn                  Bay Ridge  40.625801 -74.030621
46   Brooklyn                Bensonhurst  40.611009 -73.995180
47   Brooklyn                Sunset Park  40.645103 -74.010316
48   Brooklyn                 Greenpoint  40.730201 -73.954241
49   Brooklyn                  Gravesend  40.595260 -73.973471
50   Brooklyn             Brighton Beach  40.576825 -73.965094
51   Brooklyn             Sheepshead Bay  40.586890 -73.943186
52   Brooklyn          Manhattan Terrace  40.614433 -73.957438
53   Brooklyn                   Flatbush  40.636326 -73.958401
54   Brooklyn              Crown Heights  40.670829 -73.943291
55   Brooklyn              East Flatbush  40.641718 -73.936103
56   Brooklyn                 Kensington  40.642382 -73.980421
57   Brooklyn            Windsor Terrace  40.656946 -73.980073
58   Brooklyn           Prospect Heights  40.676822 -73.964859
59   Brooklyn                Brownsville  40.663950 -73.910235
60   Brooklyn               Williamsburg  40.707144 -73.958115
61   Brooklyn                   Bushwick  40.698116 -73.925258
62   Brooklyn         Bedford Stuyves

In [16]:
# Check to see how many neighborhoods there are.
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(bxbk_df['Borough'].unique()),
        bxbk_df.shape[0]))

The dataframe has 2 boroughs and 122 neighborhoods.


#### Use the geopy library to get the latitude and longitude values of New York City.

In [17]:
city = 'New York City, NY'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(city)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of New York City are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of New York City are 40.7127281, -74.0060152.


## Part 3. Methodology

### Exploratory Data Analysis:

1. Above dataframe which I created is for **Exploratory Data Analysis**.  
2. This dataframe contains the geographical coordinates of the two Boroughs.
3. This data will be used to get Venues data by leveraging **Fouresquare**.
4. I want to map out the neighborhoods.
5. Eventually, I will group them into 5 different clusters by using **k-means Clustering**. 

------------------------------------------------------------------------------

#### Create a Map of New York City with the neighborhoods plotted onto it.
Firstly I just wanted to map out the different neighborhoods and make sure my map looked good.  So the first thing I did was plot out the neighborhoods of the Bronx and Brooklyn using the Python Folium library. I used the latitude and longitude of each neighborhood and the map came out as below:


In [18]:
# create map of New York using latitude and longitude values
map_newyork = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(bxbk_df['Latitude'], bxbk_df['Longitude'], bxbk_df['Borough'], bxbk_df['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_newyork)  
    
map_newyork

### Utilize the Foursquare API in order to explore the neighborhoods and segment them.

#### Define Foursquare Credentials and Version.

In [19]:
CLIENT_ID = 'VW4IYSJWZOV5MEUTYKEWPEINY3TZWYE0GJHVKNUQUSJ5DVFW' # your Foursquare ID
CLIENT_SECRET = 'NORPTKWY0S0MX3LH53PPQPJ410DHEFFJMVU055BAKPE1QAH4' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Credentails:
CLIENT_ID: VW4IYSJWZOV5MEUTYKEWPEINY3TZWYE0GJHVKNUQUSJ5DVFW
CLIENT_SECRET:NORPTKWY0S0MX3LH53PPQPJ410DHEFFJMVU055BAKPE1QAH4


#### Exploring the Pelham Bay neighborhood in our dataframe.

In [20]:
bxbk_df.loc[37, 'Neighborhood']

'Pelham Bay'

In [21]:
# Get the latitude and longitude values for Pelham Bay.
hood_lat = bxbk_df.loc[37, 'Latitude']
hood_long = bxbk_df.loc[37, 'Longitude']

hood_name = bxbk_df.loc[37, 'Neighborhood']
radius=750
LIMIT=100

print('Latitude and Longitude values of {} are {}, {}'.format(hood_name,
                                                               hood_lat,
                                                               hood_long))

Latitude and Longitude values of Pelham Bay are 40.85064140940335, -73.8320737824047


### Get the top 100 venues that are located in Pelham Bay within the radius of 750 meters.
First, create the GET request URL.

In [22]:
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    hood_lat, 
    hood_long, 
    radius, 
    LIMIT)
url # display URL

'https://api.foursquare.com/v2/venues/explore?&client_id=VW4IYSJWZOV5MEUTYKEWPEINY3TZWYE0GJHVKNUQUSJ5DVFW&client_secret=NORPTKWY0S0MX3LH53PPQPJ410DHEFFJMVU055BAKPE1QAH4&v=20180605&ll=40.85064140940335,-73.8320737824047&radius=750&limit=100'

In [23]:
# Send the GET request and examie the results
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5d4513495f9e7900257d3059'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': 'Open now', 'key': 'openNow'},
    {'name': '$-$$$$', 'key': 'price'}]},
  'headerLocation': 'Pelham Bay',
  'headerFullLocation': 'Pelham Bay, Bronx',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 61,
  'suggestedBounds': {'ne': {'lat': 40.857391416153355,
    'lng': -73.82316678840145},
   'sw': {'lat': 40.84389140265335, 'lng': -73.84098077640796}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4c249bffa852c9285f52e36c',
       'name': 'Empire Bagels',
       'location': {'address': '3101 Westchester Ave',
        'lat': 40.8493916428526,
        'lng': -73.83052713165733,
        'labeledLatLngs': [

### All the information needed is in the items key. First write a function that extracts the category of venues.

In [24]:
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

### Clean the json data and structure it into a pandas dataframe.

In [25]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name          categories        lat        lng
0                   Empire Bagels          Bagel Shop  40.849392 -73.830527
1  Zeppieri & Sons Italian Bakery              Bakery  40.847119 -73.832057
2     iLoveKickboxing - Bronx, NY                 Gym  40.852871 -73.828085
3                        Key Food         Supermarket  40.850759 -73.830227
4   Giovanni's Italian Restaurant  Italian Restaurant  40.851786 -73.829786

In [26]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

61 venues were returned by Foursquare.


#### Everything looks great so far.  Now let's loo around my colleage's neighborhood, Pelham Bay.    The map is visualized below: 

In [27]:
# Plot out a map of the venues of Pelham Bay.
map_pb = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(nearby_venues['lat'], nearby_venues['lng'], nearby_venues['name'], nearby_venues['categories']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_pb)  
    
map_pb

### Explore Neighborhoods in The Bronx and Brooklyn

In [28]:
# Function to repeat the same process to all the neighborhoods in Toronto.
def getNearbyVenues(names, latitudes, longitudes, radius=750):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [29]:
# Run the above function on each neighborhood and create a new dataframe.
bxbk_venues = getNearbyVenues(names=bxbk_df['Neighborhood'],
                                 latitudes=bxbk_df['Latitude'],
                                 longitudes=bxbk_df['Longitude'])

Wakefield
Co-op City
Eastchester
Fieldston
Riverdale
Kingsbridge
Woodlawn
Norwood
Williamsbridge
Baychester
Pelham Parkway
City Island
Bedford Park
University Heights
Morris Heights
Fordham
East Tremont
West Farms
High  Bridge
Melrose
Mott Haven
Port Morris
Longwood
Hunts Point
Morrisania
Soundview
Clason Point
Throgs Neck
Country Club
Parkchester
Westchester Square
Van Nest
Morris Park
Belmont
Spuyten Duyvil
North Riverdale
Pelham Bay
Schuylerville
Edgewater Park
Castle Hill
Olinville
Pelham Gardens
Concourse
Unionport
Edenwald
Bay Ridge
Bensonhurst
Sunset Park
Greenpoint
Gravesend
Brighton Beach
Sheepshead Bay
Manhattan Terrace
Flatbush
Crown Heights
East Flatbush
Kensington
Windsor Terrace
Prospect Heights
Brownsville
Williamsburg
Bushwick
Bedford Stuyvesant
Brooklyn Heights
Cobble Hill
Carroll Gardens
Red Hook
Gowanus
Fort Greene
Park Slope
Cypress Hills
East New York
Starrett City
Canarsie
Flatlands
Mill Island
Manhattan Beach
Coney Island
Bath Beach
Borough Park
Dyker Heights
Ger

In [30]:
bxbk_venues.head()

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0    Wakefield              40.894705              -73.847201   
1    Wakefield              40.894705              -73.847201   
2    Wakefield              40.894705              -73.847201   
3    Wakefield              40.894705              -73.847201   
4    Wakefield              40.894705              -73.847201   

                                     Venue  Venue Latitude  Venue Longitude  \
0                         Lollipops Gelato       40.894123       -73.845892   
1                                 Rite Aid       40.896649       -73.844846   
2              Jackie's West Indian Bakery       40.889283       -73.843310   
3                         Carvel Ice Cream       40.890487       -73.848568   
4  Cooler Runnings Jamaican Restaurant Inc       40.898276       -73.850381   

         Venue Category  
0          Dessert Shop  
1              Pharmacy  
2  Caribbean Restaurant  
3        Ice Cream Shop  
4  Caribbean Restaurant

In [31]:
bxbk_venues.shape

(6710, 7)

### Checkout how many venues were returned for each neighborhood.

In [32]:
bxbk_venues.groupby('Neighborhood').count()

Neighborhood Latitude  Neighborhood Longitude  \
Neighborhood                                                               
Allerton                                      42                      42   
Bath Beach                                    83                      83   
Bay Ridge                                    100                     100   
Baychester                                    64                      64   
Bedford Park                                  60                      60   
Bedford Stuyvesant                            78                      78   
Belmont                                       95                      95   
Bensonhurst                                   63                      63   
Bergen Beach                                   7                       7   
Boerum Hill                                  100                     100   
Borough Park                                  29                      29   
Brighton Beach                                71                      71   
Broadway Junction                             28                      28   
Bronxdale                                     90                      90   
Brooklyn Heights                             100                     100   
Brownsville                                   42                      42   
Bushwick                                     100                     100   
Canarsie                                      19                      19   
Carroll Gardens                              100                     100   
Castle Hill                                   18                      18   
City Island                                   38                      38   
City Line                                     60                      60   
Claremont Village                             37                      37   
Clason Point                                   8                       8   
Clinton Hill                                 100                     100   
Co-op City                                    37                      37   
Cobble Hill                                  100                     100   
Concourse                                     49                      49   
Concourse Village                             55                      55   
Coney Island                                  69                      69   
Country Club                                  38                      38   
Crown Heights                                 63                      63   
Cypress Hills                                 43                      43   
Ditmas Park                                   62                      62   
Downtown                                     100                     100   
Dumbo                                        100                     100   
Dyker Heights                                 46                      46   
East Flatbush                                 25                      25   
East New York                                 31                      31   
East Tremont                                  53                      53   
East Williamsburg                            100                     100   
Eastchester                                   36                      36   
Edenwald                                      22                      22   
Edgewater Park                                42                      42   
Erasmus                                       39                      39   
Fieldston                                     14                      14   
Flatbush                                      46                      46   
Flatlands                                     33                      33   
Fordham                                       85                      85   
Fort Greene                                  100                     100   
Fort Hamilton                                 57                      57   

In [33]:
print('There are {} uniques categories.'.format(len(bxbk_venues['Venue Category'].unique())))

There are 368 uniques categories.


### Analyze Each Neighborhood.

In [34]:
# Get dummy variables.
bxbk_onehot = pd.get_dummies(bxbk_venues[['Venue Category']], prefix='', prefix_sep='')

# Add neighborhood column back to dataframe.
bxbk_onehot['Neighborhood'] = bxbk_venues['Neighborhood']

# Move neighborhood column to the first column.
fixed_columns = [bxbk_onehot.columns[-1]] + list(bxbk_onehot.columns[:-1])
bxbk_onehot = bxbk_onehot[fixed_columns]

bxbk_onehot.head()

Zoo Exhibit  Accessories Store  African Restaurant  American Restaurant  \
0            0                  0                   0                    0   
1            0                  0                   0                    0   
2            0                  0                   0                    0   
3            0                  0                   0                    0   
4            0                  0                   0                    0   

   Animal Shelter  Antique Shop  Arcade  Arepa Restaurant  \
0               0             0       0                 0   
1               0             0       0                 0   
2               0             0       0                 0   
3               0             0       0                 0   
4               0             0       0                 0   

   Argentinian Restaurant  Art Gallery  Art Museum  Arts & Crafts Store  \
0                       0            0           0                    0   
1                       0            0           0                    0   
2                       0            0           0                    0   
3                       0            0           0                    0   
4                       0            0           0                    0   

   Arts & Entertainment  Asian Restaurant  Athletics & Sports  \
0                     0                 0                   0   
1                     0                 0                   0   
2                     0                 0                   0   
3                     0                 0                   0   
4                     0                 0                   0   

   Austrian Restaurant  Auto Dealership  Auto Garage  Automotive Shop  \
0                    0                0            0                0   
1                    0                0            0                0   
2                    0                0            0                0   
3                    0                0            0                0   
4                    0                0            0                0   

   BBQ Joint  Bagel Shop  Bakery  Bank  Bar  Baseball Field  Baseball Stadium  \
0          0           0       0     0    0               0                 0   
1          0           0       0     0    0               0                 0   
2          0           0       0     0    0               0                 0   
3          0           0       0     0    0               0                 0   
4          0           0       0     0    0               0                 0   

   Basketball Court  Basketball Stadium  Bath House  Beach  Beer Bar  \
0                 0                   0           0      0         0   
1                 0                   0           0      0         0   
2                 0                   0           0      0         0   
3                 0                   0           0      0         0   
4                 0                   0           0      0         0   

   Beer Garden  Beer Store  Big Box Store  Bike Rental / Bike Share  \
0            0           0              0                         0   
1            0           0              0                         0   
2            0           0              0                         0   
3            0           0              0                         0   
4            0           0              0                         0   

   Bike Shop  Bike Trail  Bistro  Board Shop  Boat Rental  Boat or Ferry  \
0          0           0       0           0            0              0   
1          0           0       0           0            0              0   
2          0           0       0           0            0              0   
3          0           0       0           0            0              0   
4          0           0       0           0            0              0   

   Bookstore  Botanical Garden  Boutique  Bowling Alley  Boxing Gym  \
0          

In [35]:
bxbk_onehot.shape

(6710, 368)

### Group rows by neighborhood and by taking the mean of the frequency of occurance of each category.

In [36]:
bxbk_grouped = bxbk_onehot.groupby('Neighborhood').mean().reset_index()
bxbk_grouped.head()

Neighborhood  Zoo Exhibit  Accessories Store  African Restaurant  \
0      Allerton          0.0            0.00000                 0.0   
1    Bath Beach          0.0            0.00000                 0.0   
2     Bay Ridge          0.0            0.00000                 0.0   
3    Baychester          0.0            0.03125                 0.0   
4  Bedford Park          0.0            0.00000                 0.0   

   American Restaurant  Animal Shelter  Antique Shop    Arcade  \
0             0.023810             0.0           0.0  0.000000   
1             0.012048             0.0           0.0  0.000000   
2             0.020000             0.0           0.0  0.000000   
3             0.015625             0.0           0.0  0.015625   
4             0.000000             0.0           0.0  0.000000   

   Arepa Restaurant  Argentinian Restaurant  Art Gallery  Art Museum  \
0               0.0                     0.0          0.0         0.0   
1               0.0                     0.0          0.0         0.0   
2               0.0                     0.0          0.0         0.0   
3               0.0                     0.0          0.0         0.0   
4               0.0                     0.0          0.0         0.0   

   Arts & Crafts Store  Arts & Entertainment  Asian Restaurant  \
0                  0.0                   0.0          0.000000   
1                  0.0                   0.0          0.036145   
2                  0.0                   0.0          0.010000   
3                  0.0                   0.0          0.000000   
4                  0.0                   0.0          0.000000   

   Athletics & Sports  Austrian Restaurant  Auto Dealership  Auto Garage  \
0                 0.0                  0.0              0.0          0.0   
1                 0.0                  0.0              0.0          0.0   
2                 0.0                  0.0              0.0          0.0   
3                 0.0                  0.0              0.0          0.0   
4                 0.0                  0.0              0.0          0.0   

   Automotive Shop  BBQ Joint  Bagel Shop    Bakery      Bank       Bar  \
0          0.02381        0.0    0.000000  0.000000  0.047619  0.023810   
1          0.00000        0.0    0.000000  0.048193  0.060241  0.000000   
2          0.00000        0.0    0.040000  0.030000  0.000000  0.040000   
3          0.00000        0.0    0.015625  0.031250  0.031250  0.000000   
4          0.00000        0.0    0.000000  0.016667  0.033333  0.016667   

   Baseball Field  Baseball Stadium  Basketball Court  Basketball Stadium  \
0             0.0               0.0               0.0                 0.0   
1             0.0               0.0               0.0                 0.0   
2             0.0               0.0               0.0                 0.0   
3             0.0               0.0               0.0                 0.0   
4             0.0               0.0               0.0                 0.0   

   Bath House  Beach  Beer Bar  Beer Garden  Beer Store  Big Box Store  \
0         0.0    0.0       0.0     0.000000         0.0            0.0   
1         0.0    0.0       0.0     0.000000         0.0            0.0   
2         0.0    0.0       0.0     0.000000         0.0            0.0   
3         0.0    0.0       0.0     0.000000         0.0            0.0   
4         0.0    0.0       0.0     0.016667         0.0            0.0   

   Bike Rental / Bike Share  Bike Shop  Bike Trail  Bistro  Board Shop  \
0                       0.0        0.0         0.0     0.0         0.0   
1                       0.0        0.0         0.0     0.0         0.0   
2                       0.0        0.0         0.0     0.0         0.0   
3                       0.0        0.0         0.0     0.0         0.0   
4                       0.0        0.0         0.0     0.0         0.0   

   Boat Rental  Boat or Ferry  Bookstore  Botanical Garden  Boutique  \
0          0.0  

In [37]:
bxbk_grouped.shape

(122, 368)

### Print out each neighborhood with the top 5 most common venues.

I was ready to group all the categories of the different venues in each neighborhood in The Bronx and Brooklyn. There were 368 unique venue categories ranging everywhere from Adult Boutiques to Speakeasies to Zoos and of course Pizza Parlors. I then had to encode the data and get dummy variables of all the different categories. Once I did that I was able to break down the frequency of the different categories of each neighborhood. The 5 most frequent venues for my neighborhood are listed below:

In [38]:
num_top_venues = 5

for hood in bxbk_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = bxbk_grouped[bxbk_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Allerton----
               venue  freq
0        Pizza Place  0.10
1         Donut Shop  0.10
2     Sandwich Place  0.07
3  Martial Arts Dojo  0.05
4           Pharmacy  0.05


----Bath Beach----
                  venue  freq
0           Pizza Place  0.07
1                  Bank  0.06
2    Chinese Restaurant  0.06
3  Cantonese Restaurant  0.05
4                Bakery  0.05


----Bay Ridge----
                venue  freq
0                 Spa  0.07
1  Italian Restaurant  0.06
2         Pizza Place  0.05
3      Cosmetics Shop  0.04
4                 Bar  0.04


----Baychester----
              venue  freq
0       Bus Station  0.06
1        Donut Shop  0.05
2    Discount Store  0.05
3  Department Store  0.05
4          Pharmacy  0.05


----Bedford Park----
            venue  freq
0   Deli / Bodega  0.08
1     Pizza Place  0.08
2     Supermarket  0.07
3           Diner  0.07
4  Sandwich Place  0.05


----Bedford Stuyvesant----
           venue  freq
0    Pizza Place  0.09
1    Coffee S

### Put the most common venues into a dataframe.
Now I was almost ready to start clustering the neighborhoods. To make things easier I only wanted to use the top 10 venues in each neighborhood for comparison. I created a new dataframe with the neighborhoods and the top 10 venues and the top 5 rows  which looked like this:

In [39]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [40]:
# Create the new dataframe and display the top 10 venues for each neighborhood.
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = bxbk_grouped['Neighborhood']

for ind in np.arange(bxbk_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(bxbk_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
0      Allerton            Donut Shop           Pizza Place   
1    Bath Beach           Pizza Place                  Bank   
2     Bay Ridge                   Spa    Italian Restaurant   
3    Baychester           Bus Station        Discount Store   
4  Bedford Park         Deli / Bodega           Pizza Place   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0        Sandwich Place     Martial Arts Dojo  Fast Food Restaurant   
1    Chinese Restaurant                Bakery  Cantonese Restaurant   
2           Pizza Place      Greek Restaurant        Cosmetics Shop   
3      Department Store              Pharmacy            Donut Shop   
4                 Diner           Supermarket        Sandwich Place   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0              Pharmacy           Supermarket                  Bank   
1       Supplement Shop            Donut Shop       Bubble Tea Shop   
2                   Bar            Bagel Shop                Bakery   
3                Bakery  Fast Food Restaurant                  Bank   
4                Garden    Chinese Restaurant                  Bank   

  9th Most Common Venue 10th Most Common Venue  
0        Clothing Store     Seafood Restaurant  
1    Italian Restaurant            Coffee Shop  
2    Chinese Restaurant                  Diner  
3        Clothing Store     Spanish Restaurant  
4      Botanical Garden             Donut Shop

### Cluster Neighborhoods.
#### Run _k_-means to cluster the neighborhoods into 5 clusters.
Once that was done I could run a k-means clustering algorithm on the data and assign a cluster to every neighborhood that was similar to one another. From there I was able to look at my results and compare the neighborhoods. 

In [41]:
# Set the number of clusters
kclusters = 5

bxbk_clustering = bxbk_grouped.drop('Neighborhood', 1)

# Run the k-means clustering.
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(bxbk_clustering)

# Check the cluster labels generated for each row in the dataframe.
kmeans.labels_[0:10]

array([2, 2, 4, 2, 2, 4, 4, 2, 1, 4], dtype=int32)

### Create a new dataframe that includes the clusters as well as the top 10 venues for each neighborhood.

In [42]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

bxbk_merged = bxbk_df

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
bxbk_merged = bxbk_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

bxbk_merged.head() 

Borough Neighborhood   Latitude  Longitude  Cluster Labels  \
0   Bronx    Wakefield  40.894705 -73.847201               2   
1   Bronx   Co-op City  40.874294 -73.829939               2   
2   Bronx  Eastchester  40.887556 -73.827806               2   
3   Bronx    Fieldston  40.895437 -73.905643               0   
4   Bronx    Riverdale  40.890834 -73.912585               4   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0           Supermarket            Donut Shop   Fried Chicken Joint   
1           Bus Station           Pizza Place           Supermarket   
2  Caribbean Restaurant  Fast Food Restaurant         Deli / Bodega   
3           Bus Station                 Plaza            Playground   
4                  Bank           Pizza Place            Playground   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0              Pharmacy           Gas Station  Caribbean Restaurant   
1  Fast Food Restaurant        Mattress Store      Department Store   
2         Women's Store                 Diner       Supplement Shop   
3    Athletics & Sports           Art Gallery      Basketball Court   
4    Mexican Restaurant                   Bar   Japanese Restaurant   

  7th Most Common Venue   8th Most Common Venue 9th Most Common Venue  \
0      Asian Restaurant    Fast Food Restaurant                 Track   
1  Caribbean Restaurant                   Trail         Grocery Store   
2           Supermarket  Furniture / Home Store         Bowling Alley   
3           Coffee Shop            Home Service           Pizza Place   
4                 Diner                   Plaza            Bagel Shop   

  10th Most Common Venue  
0         Sandwich Place  
1     Chinese Restaurant  
2         Breakfast Spot  
3                  River  
4      Mobile Phone Shop

### Finally visualize the resulting clusters on the map.

In [43]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(bxbk_merged['Latitude'], bxbk_merged['Longitude'], bxbk_merged['Neighborhood'], bxbk_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## Part 4. Result
#### Examine the Clusters.

### Cluster 1

In [44]:
bxbk_merged.loc[bxbk_merged['Cluster Labels'] == 0, bxbk_merged.columns[[1] + list(range(5, bxbk_merged.shape[1]))]]

Neighborhood 1st Most Common Venue       2nd Most Common Venue  \
3           Fieldston           Bus Station                       Plaza   
12        City Island       Harbor / Marina               Deli / Bodega   
24        Hunts Point                  Park                 Pizza Place   
27       Clason Point                  Park                    Bus Stop   
35     Spuyten Duyvil                  Park                        Bank   
51     Brighton Beach                 Beach                      Bakery   
52     Sheepshead Bay         Grocery Store              Sandwich Place   
64   Brooklyn Heights                  Park                 Yoga Studio   
67           Red Hook         Boat or Ferry                        Park   
76        Mill Island              Gym Pool  Tourist Information Center   
77    Manhattan Beach       Harbor / Marina               Boat or Ferry   
261   Paerdegat Basin      Asian Restaurant          Chinese Restaurant   
279      Fulton Ferry                  Park                 Coffee Shop   
280      Vinegar Hill                  Park                   Bookstore   
283             Dumbo                  Park                 Coffee Shop   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
3              Playground    Athletics & Sports           Art Gallery   
12     Italian Restaurant    Seafood Restaurant   American Restaurant   
24          Moving Target           Supermarket           Fish Market   
27         Scenic Lookout         Boat or Ferry         Grocery Store   
35               Pharmacy            Donut Shop          Tennis Court   
51             Restaurant           Supermarket         Grocery Store   
52             Restaurant    Italian Restaurant              Pharmacy   
64     Italian Restaurant           Coffee Shop        Ice Cream Shop   
67    American Restaurant    Seafood Restaurant           Art Gallery   
76        Harbor / Marina                  Pool    Seafood Restaurant   
77           Dessert Shop    Turkish Restaurant            Bagel Shop   
261  Gym / Fitness Center                  Food                   Gym   
279   American Restaurant                  Café        Scenic Lookout   
280                  Café                   Gym           Coffee Shop   
283        Scenic Lookout   American Restaurant                  Café   

          6th Most Common Venue 7th Most Common Venue  \
3              Basketball Court           Coffee Shop   
12                Boat or Ferry        Ice Cream Shop   
24                     Bus Stop          Gourmet Shop   
27    South American Restaurant                   Zoo   
35                   Bagel Shop         Train Station   
51           Russian Restaurant           Coffee Shop   
52                         Park    Seafood Restaurant   
64                    Wine Shop           Pizza Place   
67                    Wine Shop           Coffee Shop   
76   Construction & Landscaping          Burger Joint   
77                   Restaurant              Bus Stop   
261               Moving Target         Grocery Store   
279          Seafood Restaurant                Bakery   
280                  Boxing Gym             Wine Shop   
283                         Bar                Bakery   

           8th Most Common Venue   9th Most Common Venue  \
3                   Home Service             Pizza Place   
12                           Bar  Thrift / Vintage Store   
24                Farmers Market           Grocery Store   
27             Fish & Chips Shop                    Farm   
35                          Café              Board Shop   
51   Eastern European Restaurant    Fast Food Restaurant   
52            Chinese Restaurant      Russian Restaurant   
64       New American Restaurant    Gym / Fitness Center   
67                          Café              Bagel Shop   
76                    Playground               BBQ Joint   
77                          Café        Department Store   
261   

## Cluster 2

In [45]:
bxbk_merged.loc[bxbk_merged['Cluster Labels'] == 1, bxbk_merged.columns[[1] + list(range(5, bxbk_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
91  Bergen Beach       Harbor / Marina            Donut Shop   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
91            Playground                   Gym    Athletics & Sports   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
91        Baseball Field            Food Court     Food & Drink Shop   

   9th Most Common Venue 10th Most Common Venue  
91    Falafel Restaurant    Fried Chicken Joint

## Cluster 3

In [46]:
bxbk_merged.loc[bxbk_merged['Cluster Labels'] == 2, bxbk_merged.columns[[1] + list(range(5, bxbk_merged.shape[1]))]]

Neighborhood         1st Most Common Venue 2nd Most Common Venue  \
0              Wakefield                   Supermarket            Donut Shop   
1             Co-op City                   Bus Station           Pizza Place   
2            Eastchester          Caribbean Restaurant  Fast Food Restaurant   
8                Norwood                   Pizza Place    Spanish Restaurant   
9         Williamsbridge          Caribbean Restaurant           Pizza Place   
10            Baychester                   Bus Station        Discount Store   
11        Pelham Parkway                   Pizza Place        Sandwich Place   
13          Bedford Park                 Deli / Bodega           Pizza Place   
14    University Heights                   Pizza Place   Fried Chicken Joint   
15        Morris Heights            Spanish Restaurant            Food Truck   
16               Fordham            Spanish Restaurant            Donut Shop   
17          East Tremont                   Pizza Place                  Bank   
18            West Farms                          Park  Fast Food Restaurant   
19          High  Bridge                   Pizza Place                  Park   
20               Melrose                   Pizza Place           Bus Station   
21            Mott Haven                           Gym            Donut Shop   
22           Port Morris                           Gym        Baseball Field   
23              Longwood          Fast Food Restaurant         Deli / Bodega   
25            Morrisania                Discount Store            Donut Shop   
26             Soundview          Fast Food Restaurant         Deli / Bodega   
30           Parkchester                   Supermarket           Pizza Place   
31    Westchester Square          Fast Food Restaurant            Donut Shop   
32              Van Nest                   Pizza Place         Deli / Bodega   
33           Morris Park                   Pizza Place        Sandwich Place   
37            Pelham Bay            Italian Restaurant                  Bank   
38         Schuylerville                   Pizza Place                   Bar   
40           Castle Hill                   Pizza Place        Baseball Field   
41             Olinville                 Deli / Bodega           Pizza Place   
42        Pelham Gardens            Chinese Restaurant           Bus Station   
43             Concourse                Ice Cream Shop  Fast Food Restaurant   
44             Unionport                   Pizza Place          Dance Studio   
45              Edenwald          Caribbean Restaurant           Supermarket   
47           Bensonhurst            Chinese Restaurant            Donut Shop   
48           Sunset Park                   Pizza Place                Bakery   
50             Gravesend                   Pizza Place    Chinese Restaurant   
53     Manhattan Terrace                   Pizza Place            Donut Shop   
56         East Flatbush          Caribbean Restaurant         Deli / Bodega   
57            Kensington                 Deli / Bodega           Pizza Place   
60           Brownsville                   Pizza Place        Discount Store   
71         Cypress Hills                    Donut Shop  Fast Food Restaurant   
72         East New York                 Deli / Bodega  Fast Food Restaurant   
73         Starrett City              Department Store           Supermarket   
74              Canarsie          Caribbean Restaurant                  Bank   
75             Flatlands                 Deli / Bodega              Pharmacy   
78          Coney Island  Theme Park Ride / Attraction                 Beach   
79            Bath Beach                   Pizza Place                  Bank   
80          Borough Park                          Bank           Pizza Place   
83           Marine Park            Athletics & Sports              Pharmacy   
89            Ocean Hill                 Deli / Bodega        Discount Store   
90          

## Cluster 4

In [47]:
bxbk_merged.loc[bxbk_merged['Cluster Labels'] == 3, bxbk_merged.columns[[1] + list(range(5, bxbk_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
85     Sea Gate                 Beach                  Park   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
85           Supermarket                   Spa          Home Service   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
85     Fish & Chips Shop    Falafel Restaurant                  Farm   

   9th Most Common Venue 10th Most Common Venue  
85        Farmers Market   Fast Food Restaurant

## Cluster 5

In [48]:
bxbk_merged.loc[bxbk_merged['Cluster Labels'] == 4, bxbk_merged.columns[[1] + list(range(5, bxbk_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
4                    Riverdale                  Bank           Pizza Place   
5                  Kingsbridge           Pizza Place        Sandwich Place   
7                     Woodlawn                   Bar         Deli / Bodega   
28                 Throgs Neck    Italian Restaurant         Deli / Bodega   
29                Country Club                Bakery        Sandwich Place   
34                     Belmont    Italian Restaurant           Pizza Place   
36             North Riverdale           Pizza Place           Bus Station   
39              Edgewater Park         Deli / Bodega    Italian Restaurant   
46                   Bay Ridge                   Spa    Italian Restaurant   
49                  Greenpoint                   Bar           Coffee Shop   
54                    Flatbush  Caribbean Restaurant           Pizza Place   
55               Crown Heights           Pizza Place                  Café   
58             Windsor Terrace         Deli / Bodega    Italian Restaurant   
59            Prospect Heights                   Bar    Mexican Restaurant   
61                Williamsburg                   Bar           Pizza Place   
62                    Bushwick                   Bar    Mexican Restaurant   
63          Bedford Stuyvesant           Pizza Place           Coffee Shop   
65                 Cobble Hill    Italian Restaurant                   Bar   
66             Carroll Gardens    Italian Restaurant           Coffee Shop   
68                     Gowanus    Italian Restaurant                   Bar   
69                 Fort Greene    Italian Restaurant           Coffee Shop   
70                  Park Slope           Coffee Shop   American Restaurant   
81               Dyker Heights        Cosmetics Shop           Flower Shop   
82             Gerritsen Beach         Deli / Bodega    Seafood Restaurant   
84                Clinton Hill    Italian Restaurant           Coffee Shop   
86                    Downtown          Burger Joint           Pizza Place   
87                 Boerum Hill                   Bar           Coffee Shop   
88   Prospect Lefferts Gardens  Caribbean Restaurant           Pizza Place   
92                     Midwood           Pizza Place            Bagel Shop   
95           East Williamsburg           Coffee Shop                   Bar   
96                  North Side           Yoga Studio           Coffee Shop   
97                  South Side           Coffee Shop                   Bar   
99               Fort Hamilton           Pizza Place   Japanese Restaurant   
221                Ditmas Park    Mexican Restaurant  Caribbean Restaurant   

    3rd Most Common Venue 4th Most Common Venue      5th Most Common Venue  \
4              Playground    Mexican Restaurant                        Bar   
5                     Bar    Mexican Restaurant                      Diner   
7                     Pub           Pizza Place         Italian Restaurant   
28            Coffee Shop           Pizza Place       Fast Food Restaurant   
29                   Bank    Italian Restaurant         Mexican Restaurant   
34                 Bakery         Deli / Bodega             Sandwich Place   
36                   Pool        Sandwich Place           Sushi Restaurant   
39            Pizza Place      Asian Restaurant                        Bar   
46            Pizza Place      Greek Restaurant             Cosmetics Shop   
49           Cocktail Bar           Pizza Place                       Café   
54                 Lounge    Mexican Restaurant              Deli / Bodega   
55   Caribbean Restaurant         Grocery Store                Coffee Shop   
58         Ice Cream Shop                  Park                       Café   
59              Wine Shop           Coffee Shop                Yoga Studio   
61            Coffee Shop   American Restaurant                Yoga Studio   
62            Coffee Shop              

## Part 5. Conclusion

- My colleague's neighborhood, Pelham Bay was located in the largest cluster (**Cluster 3**) of all of them. Apparently this neighborhood isn't too unique. 

- This neighborhood is much more like most other Bronx neighborhoods and southern and eastern Brooklyn neighborhoods. 

- I recommended to move to the north eastern part of Brooklyn (the green cluster). Apparently it's not as similar to current neighborhood. However, the gentrification of those Brooklyn neighborhoods and hipster influence on this decision. 

##### Some further examination to be done:
- Add more clusters.
- Look at features such as crime rate, demographics, and median age of each neighborhood. 
- Use the top 20 or 30 venues next time instead of just the top 10. 